src/replication/ng/flow/flow.go

```diff
@@ -0,0 +1,206 @@
+	}
+	srcAdapter, err := srcFactory(srcRegistry)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create adapter for source registry %s: %v", srcRegistry.URL, err)
+	}
+	f.srcAdapter = srcAdapter
+
+	// create the destination registry adapter
+	dstFactory, err := adapter.GetFactory(dstRegistry.Type)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get adapter factory for registry type %s: %v", dstRegistry.Type, err)
+	}
+	dstAdapter, err := dstFactory(dstRegistry)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create adapter for destination registry %s: %v", dstRegistry.URL, err)
+	}
+	f.dstAdapter = dstAdapter
+
+	return f, nil
+}
+
+func (f *flow) createExecution() (int64, error) {
+	id, err := f.executionMgr.Create(&model.Execution{
+		PolicyID:  f.policy.ID,
+		Status:    model.ExecutionStatusInProgress,
+		StartTime: time.Now(),
+	})
+	f.executionID = id
+	log.Debugf("an execution record for replication based on the policy %d created: %d", f.policy.ID, id)
+	return id, err
+}
+
+func (f *flow) fetchResources() error {
+	resources, err := f.srcAdapter.FetchResources(f.policy.SrcNamespaces, f.policy.Filters)
+	f.resources = resources
+	if err != nil {
+		f.markExecutionFailure(err)
+		return err
+	}
+
+	log.Debugf("resources for the execution %d fetched from the source registry", f.executionID)
+	return nil
+}
+
+func (f *flow) createNamespace() error {
+	// merge the metadata of all source namespaces
+	metadata := map[string]interface{}{}
+	for _, srcNamespace := range f.policy.SrcNamespaces {
+		namespace, err := f.srcAdapter.GetNamespace(srcNamespace)
+		if err != nil {
+			f.markExecutionFailure(err)
+			return err
+		}
+		for key, value := range namespace.Metadata {
+			metadata[namespace.Name+":"+key] = value
+		}
+	}
+
+	if err := f.dstAdapter.CreateNamespace(&model.Namespace{
+		Name:     f.policy.DestNamespace,
+		Metadata: metadata,
+	}); err != nil {
+		f.markExecutionFailure(err)
+		return err
+	}
+
+	log.Debugf("namespace %s for the execution %d created on the destination registry", f.policy.DestNamespace, f.executionID)
+	return nil
+}
+
+func (f *flow) schedule() error {
+	dstResources := []*model.Resource{}
+	for _, srcResource := range f.resources {
+		dstResource := &model.Resource{
+			Type: srcResource.Type,
+			Metadata: &model.ResourceMetadata{
+				Name:      srcResource.Metadata.Name,
+				Namespace: f.policy.DestNamespace,
+				Vtags:     srcResource.Metadata.Vtags,
+			},
+			Registry:     f.dstRegistry,
+			ExtendedInfo: srcResource.ExtendedInfo,
+			Deleted:      srcResource.Deleted,
+			Override:     f.policy.Override,
+		}
+		dstResources = append(dstResources, dstResource)
+	}
+
+	tasks, err := f.scheduler.Schedule(f.resources, dstResources)
+	if err != nil {
+		f.markExecutionFailure(err)
+		return err
+	}
+
+	allFailed := true
+	for _, task := range tasks {
+		if task.Status != model.TaskStatusFailed {
```

Comment:
when a task will be set to failed even before creating?  does action here match with the one in scheduler?

Reply:
The new changes that we talked about will be pushed in later PR.